### PR TITLE
Let user specify a set of vectors to use for training in `IVF_FLAT`

### PR DESCRIPTION
**Pull request by jparismorgan**
### What
First part of [Support random sampling for vector search training](https://app.shortcut.com/tiledb-inc/story/38178/support-random-sampling-for-vector-search-training). Here we add the ability for a user to point to a set of vectors to use for training. These vectors can either be local numpy arrays or already be stored at a provided URI.

### TODO
In a follow-up we will add the ability to:
1. Store this URI in metadata and use it for training during re-ingestion.
2. Add the ability for a user to specify they want to select a random set of N vectors from the input data, instead of what we offer today, i.e. for training to either a. use the first N vectors from the source data b. use the provided set of vectors we add here.

### Open questions
1. Is the approach I used to add `TRAINING_INPUT_VECTORS_ARRAY_NAME` correct? Did I miss anything there?
2. Is there any cleanup / lifetime management code I need to add around `TRAINING_INPUT_VECTORS_ARRAY_NAME`? I didn't see any, but asking just in case.

### Testing
* Added new unit tests for providing a set of vectors to use for training as a TileDB array, a numpy array, and a FVECS array
* Manually printed out the `sample_vectors` we end up using and can see they are the same as the passed in vectors - here are the results for a TileDB array:
```python
test/test_ingestion.py [test_ingestion@test_ingest_with_training_source_uri_tdb] ====================================
[test_ingestion] data (4, 5) float32 
 [[1.  2.  3.  4.  5. ]
 [1.1 2.1 3.1 4.1 5.1]
 [1.2 2.2 3.2 4.2 5.2]
 [1.3 2.3 3.3 4.3 5.3]]
[test_ingestion] training_data (2, 5) float32 
 [[1.1 2.1 3.1 4.1 5.1]
 [1.2 2.2 3.2 4.2 5.2]]
training_data.tdb ArraySchema(
  domain=Domain(*[
    Dim(name='__dim_0', domain=(0, 1), tile=1, dtype='int32', filters=FilterList([ZstdFilter(level=-1), ])),
    Dim(name='__dim_1', domain=(0, 4), tile=3, dtype='int32', filters=FilterList([ZstdFilter(level=-1), ])),
  ]),
  attrs=[
    Attr(name='values', dtype='float32', var=False, nullable=False, enum_label=None),
  ],
  cell_order='col-major',
  tile_order='col-major',
  sparse=False,
)

[ivf_flat_index@create]
[ingestion@centralised_kmeans] training_sample_size:  5 training_source_uri:  /private/var/folders/jb/5gq49wh97wn0j7hj6zfn9pzh0000gn/T/pytest-of-parismorgan/pytest-889/test_ingest_with_training_sour0/dataset/training_data.tdb training_source_type:  None
[ingestion@centralised_kmeans] training_in_size:  5 training_dimensions:  2 training_vector_type:  float32
[ingestion@centralised_kmeans] sample_vectors (5, 2) [[1.1 1.2]
 [2.1 2.2]
 [3.1 3.2]
 [4.1 4.2]
 [5.1 5.2]]
```

### CODE REVIEW

File: apis/python/src/tiledb/vector_search/ingestion.py [hunk 1]
```diff
@@ -765,45 +808,61 @@ def centralised_kmeans(
...
                         n_init=n_init,
+                        random_state=0,
                     )
                     km.fit_predict(sample_vectors)
                     centroids = np.transpose(np.array(km.cluster_centers_))
                 else:
                     centroids = kmeans_fit(partitions, init, max_iter, verbose, n_init, array_to_matrix(np.transpose(sample_vectors)))
                     centroids = np.array(centroids) # TODO: why is this here?
             else:
+                # NOTE(paris): Should we instead take the first training_sample_size vectors and then fill in random for the rest?
```
Comment (NikolaosPapailiou) [inline apis/python/src/tiledb/vector_search/ingestion.py:862]: Lets raise an exception if `training_sample_size < partitions`

File: apis/python/src/tiledb/vector_search/ivf_flat_index.py [hunk 2]
```diff
@@ -454,6 +454,7 @@ def create(
     storage_version: str = STORAGE_VERSION,
     **kwargs,
 ) -> IVFFlatIndex:
+    print('[ivf_flat_index@create]')
```
Comment (NikolaosPapailiou) [inline apis/python/src/tiledb/vector_search/ivf_flat_index.py:457]: nit: remove

File: apis/python/test/common.py [hunk 3]
```diff
@@ -193,16 +214,16 @@ def create_random_dataset_u8(nb, d, nq, k, path):
     return data
 
 
-def create_schema():
+def create_schema(dimension0DomainMax, dimension1DomainMax):
```
Comment (NikolaosPapailiou) [inline apis/python/test/common.py:217]: nit: use variable names with underscores and lowercase

**Review by NikolaosPapailiou — APPROVED** [review]

File: apis/python/src/tiledb/vector_search/ingestion.py (hunk 1 above)
Comment (jparismorgan) [reply apis/python/src/tiledb/vector_search/ingestion.py]: Looks like we actually would throw in one of our unit tests with that change: https://github.com/TileDB-Inc/TileDB-Vector-Search/actions/runs/7248899669/job/19745743298?pr=169#step:6:374

I'll change this into a TODO and investigate in a follow-up.

File: apis/python/src/tiledb/vector_search/ivf_flat_index.py (hunk 2 above)
Comment (jparismorgan) [reply apis/python/src/tiledb/vector_search/ivf_flat_index.py]: Thanks, `git add .` error, done.

File: apis/python/test/common.py (hunk 3 above)
Comment (jparismorgan) [reply apis/python/test/common.py]: Too much C++ 😅 Done.

**Pull request merged**